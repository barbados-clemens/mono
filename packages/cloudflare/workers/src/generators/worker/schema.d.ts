export interface CloudflareWorkersGeneratorSchema {
  name: string;
  tags?: string;
  directory?: string;
}
