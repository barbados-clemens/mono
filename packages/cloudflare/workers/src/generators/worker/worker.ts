import {
  addProjectConfiguration,
  formatFiles,
  generateFiles,
  GeneratorCallback,
  getWorkspaceLayout,
  joinPathFragments,
  names,
  offsetFromRoot,
  Tree,
} from '@nrwl/devkit';
import { Linter, lintProjectGenerator } from '@nrwl/linter';
import { getRootTsConfigPath } from 'nx/src/utils/typescript';
import * as path from 'path';
import { workerInit } from '../init/init';
import { CloudflareWorkersGeneratorSchema } from './schema';

interface NormalizedSchema extends CloudflareWorkersGeneratorSchema {
  projectName: string;
  projectRoot: string;
  projectDirectory: string;
  parsedTags: string[];
}

function normalizeOptions(
  tree: Tree,
  options: CloudflareWorkersGeneratorSchema
): NormalizedSchema {
  const name = names(options.name).fileName;
  const projectDirectory = options.directory
    ? `${names(options.directory).fileName}/${name}`
    : name;
  const projectName = projectDirectory.replace(new RegExp('/', 'g'), '-');
  const projectRoot = `${getWorkspaceLayout(tree).appsDir}/${projectDirectory}`;
  const parsedTags = options.tags
    ? options.tags.split(',').map((s) => s.trim())
    : [];

  return {
    ...options,
    projectName,
    projectRoot,
    projectDirectory,
    parsedTags,
  };
}

function addFiles(tree: Tree, options: NormalizedSchema) {
  const n = names(options.name);
  const offset = offsetFromRoot(options.projectRoot);
  const { npmScope } = getWorkspaceLayout(tree);
  const templateOptions = {
    ...options,
    ...n,
    rootTsConfigPath: getRootTsConfigPath(),
    importPath: `@${npmScope}/${options.projectDirectory}`,
    compatibilityDate: new Date().toISOString().split('T')[0],
    offsetFromRoot: offset,
    tmpl: '',
  };
  generateFiles(
    tree,
    path.join(__dirname, 'files'),
    options.projectRoot,
    templateOptions
  );
}

function addProject(tree: Tree, options: NormalizedSchema) {
  addProjectConfiguration(
    tree,
    options.projectName,
    {
      root: options.projectRoot,
      tags: options.parsedTags,
      projectType: 'application',
      sourceRoot: `${options.projectRoot}/src`,
      targets: {
        // wrangler cli can handle ts code
        // build: {
        //   executor: '@nrwl/js:tsc',
        //   outputs: ['{options.outputPath}'],
        //   options: {
        //     outputPath: `dist/${options.projectRoot}`,
        //     main: `${options.projectRoot}/src/index.ts`,
        //     tsConfig: `${options.projectRoot}/tsconfig.app.json`,
        //     assets: [`${options.projectRoot}/*.md`],
        //   },
        // },
        serve: {
          executor: '@nrwl/workspace:run-commands',
          options: {
            commands: [
              'npx wrangler dev src/index.ts --tsconfig=tsconfig.app.json --env=dev',
            ],
            cwd: options.projectRoot,
          },
          configurations: {
            production: {
              commands: [
                'npx wrangler dev src/index.ts --tsconfig=tsconfig.app.json --env=production',
              ],
            },
          },
        },
        publish: {
          executor: '@nrwl/workspace:run-commands',
          options: {
            commands: [
              `npx wrangler publish src/index.ts --tsconfig=tsconfig.app.json --env=dev`,
            ],
            cwd: options.projectRoot,
          },
          configurations: {
            production: {
              commands: [
                `npx wrangler publish src/index.ts --tsconfig=tsconfig.app.json --env=production`,
              ],
            },
          },
        },
      },
    },
    true
  );
}

export function addLint(
  tree: Tree,
  options: NormalizedSchema
): Promise<GeneratorCallback> {
  return lintProjectGenerator(tree, {
    project: options.projectName,
    linter: Linter.EsLint,
    skipFormat: true,
    tsConfigPaths: [
      joinPathFragments(options.projectRoot, 'tsconfig.app.json'),
    ],
    eslintFilePatterns: [`${options.projectRoot}/**/*.ts`],
  });
}

export async function workerApp(
  tree: Tree,
  options: CloudflareWorkersGeneratorSchema
) {
  const normalizedOptions = normalizeOptions(tree, options);
  const cfWorkerInitTask = workerInit(tree);

  addFiles(tree, normalizedOptions);
  addProject(tree, normalizedOptions);
  const lintTask = await addLint(tree, normalizedOptions);

  return () => {
    lintTask();
    formatFiles(tree);
    cfWorkerInitTask();
  };
}

export default workerApp;

interface TsConfig {
  include?: string[];
  compilerOptions?: Record<string, unknown>;
}
