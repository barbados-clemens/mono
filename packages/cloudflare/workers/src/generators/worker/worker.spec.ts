import { readProjectConfiguration, Tree } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { workerApp } from './worker';

describe('cloudflare-workers generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace(2);
  });

  it('should generate files', async () => {
    await workerApp(tree, {
      name: 'my-worker-app',
    });

    const config = readProjectConfiguration(tree, 'my-worker-app');
    const wranglerConfig = tree.read(
      'apps/my-worker-app/wrangler.toml',
      'utf-8'
    );
    expect(wranglerConfig).toContain('name = "my-worker-app"');
    expect(wranglerConfig).toContain('compatibility_date = "');
    expect(tree.exists('apps/my-worker-app/src/index.ts')).toBeTruthy();
    expect(
      tree.exists('apps/my-worker-app/src/request-handler.ts')
    ).toBeTruthy();
    expect(config).toMatchSnapshot();
  });
});
