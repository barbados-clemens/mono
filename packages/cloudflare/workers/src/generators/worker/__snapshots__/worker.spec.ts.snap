// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cloudflare-workers generator should generate files 1`] = `
Object {
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "projectType": "application",
  "root": "apps/my-worker-app",
  "sourceRoot": "apps/my-worker-app/src",
  "tags": Array [],
  "targets": Object {
    "lint": Object {
      "executor": "@nrwl/linter:eslint",
      "options": Object {
        "lintFilePatterns": Array [
          "apps/my-worker-app/**/*.ts",
        ],
      },
      "outputs": Array [
        "{options.outputFile}",
      ],
    },
    "publish": Object {
      "configurations": Object {
        "production": Object {
          "commands": Array [
            "npx wrangler publish src/index.ts --tsconfig=tsconfig.app.json --env=production",
          ],
        },
      },
      "executor": "@nrwl/workspace:run-commands",
      "options": Object {
        "commands": Array [
          "npx wrangler publish src/index.ts --tsconfig=tsconfig.app.json --env=dev",
        ],
        "cwd": "apps/my-worker-app",
      },
    },
    "serve": Object {
      "configurations": Object {
        "production": Object {
          "commands": Array [
            "npx wrangler dev src/index.ts --tsconfig=tsconfig.app.json --env=production",
          ],
        },
      },
      "executor": "@nrwl/workspace:run-commands",
      "options": Object {
        "commands": Array [
          "npx wrangler dev src/index.ts --tsconfig=tsconfig.app.json --env=dev",
        ],
        "cwd": "apps/my-worker-app",
      },
    },
  },
}
`;
